pkg/http/listener.go

```diff
@@ -128,65 +141,51 @@ func (listener *httpListener) start() {
 			return
 		}
 
-		// Return bufconn if read data is a valid HTTP method.
-		tokens := strings.SplitN(string(data), " ", 2)
-		if isHTTPMethod(tokens[0]) {
-			if listener.tlsConfig == nil {
-				send(acceptResult{bufconn, nil}, doneCh)
-			} else {
+		if !guessTLSProtocol(data) {
+			// Return bufconn if read data is a valid HTTP method.
+			if listener.tlsConfig != nil {
 				// As TLS is configured and we got plain text HTTP request,
 				// return 403 (forbidden) error.
 				bufconn.Write(sslRequiredErrMsg)
 				bufconn.Close()
+				return
 			}
+			send(acceptResult{bufconn, nil}, doneCh)
 			return
 		}
 
-		if listener.tlsConfig != nil {
-			// As the listener is configured with TLS, try to do TLS handshake, drop the connection if it fails.
-			tlsConn := tls.Server(bufconn, listener.tlsConfig)
-			if err := tlsConn.Handshake(); err != nil {
-				if listener.errorLogFunc != nil {
-					listener.errorLogFunc(err,
-						"TLS handshake failed with new connection %s at server %s",
-						bufconn.RemoteAddr(), bufconn.LocalAddr())
-				}
-				bufconn.Close()
-				return
-			}
-
-			// Check whether the connection contains HTTP request or not.
-			bufconn = newBufConn(tlsConn, listener.readTimeout, listener.writeTimeout,
-				listener.updateBytesReadFunc, listener.updateBytesWrittenFunc)
-
-			// Peek bytes of maximum length of all HTTP methods.
-			data, err := bufconn.Peek(methodMaxLen)
-			if err != nil {
-				if listener.errorLogFunc != nil {
-					listener.errorLogFunc(err,
-						"Error in reading from new TLS connection %s at server %s",
-						bufconn.RemoteAddr(), bufconn.LocalAddr())
-				}
-				bufconn.Close()
-				return
-			}
-
-			// Return bufconn if read data is a valid HTTP method.
-			tokens := strings.SplitN(string(data), " ", 2)
-			if isHTTPMethod(tokens[0]) {
-				send(acceptResult{bufconn, nil}, doneCh)
-				return
+		if listener.tlsConfig == nil {
+			// Client is TLS and server is not TLS configured.
+			// Can't send any error here, since client will not
+			// be able to interpret it on an attempted TLS conn.
+			if listener.errorLogFunc != nil {
+				listener.errorLogFunc(errors.New(""),
+					"Server %s not configured to support TLS request from %s",
+					bufconn.LocalAddr(), bufconn.RemoteAddr())
 			}
+			bufconn.Close()
+			return
 		}
 
-		if listener.errorLogFunc != nil {
-			listener.errorLogFunc(errors.New("junk message"),
-				"Received non-HTTP message from new connection %s at server %s",
-				bufconn.RemoteAddr(), bufconn.LocalAddr())
+		// As the listener is configured with TLS, try to do TLS
+		// handshake, drop the connection if it fails.
+		tlsConn := tls.Server(bufconn, listener.tlsConfig)
+		if err = tlsConn.Handshake(); err != nil {
+			if listener.errorLogFunc != nil {
+				listener.errorLogFunc(err,
+					"TLS handshake failed with new connection %s at server %s",
+					bufconn.RemoteAddr(), bufconn.LocalAddr())
+			}
+			bufconn.Close()
+			return
```

Comment:
@balamurugana we are okay to receive a TLS connection which is not HTTP. why would you want to check that? can you show a symptom on what happens?